core/arns/src/test/java/software/amazon/awssdk/arns/ArnTest.java

```diff
@@ -311,4 +317,78 @@ public void invalidArnWithoutAccountId_ThrowsIllegalArgumentException() {
+            Arguments.of("Resource type and resource and qualifier slash",
+                         "arn:aws:s3:us-east-1:12345678910:bucket/foobar/1"),
+            Arguments.of("Without region", "arn:aws:s3::123456789012:myresource"),
+            Arguments.of("Without accountId", "arn:aws:s3:us-east-1::myresource"),
+            Arguments.of("Resource with dots", "arn:aws:s3:us-east-1:12345678910:myresource:foobar.1")
+        );
+    }
+
+    private static Stream<Arguments> invalidArnTestCases() {
+        return Stream.of(
+            Arguments.of("Without resource", "arn:aws:s3:us-east-1:12345678910:"),
+            Arguments.of("Invalid arn", "arn:aws:"),
+            Arguments.of("Doesn't start with arn", "fakearn:aws:"),
+            Arguments.of("Invalid without partition", "arn:"),
+            Arguments.of("Invalid without service", "arn:aws:"),
+            Arguments.of("Invalid without region", "arn:aws:s3:"),
+            Arguments.of("Invalid without accountId", "arn:aws:s3:us-east-1:"),
+            Arguments.of("Null Arn", null)
+        );
+    }
+
+    private static Stream<Arguments> exceptionThrowingArnTestCases() {
+        return Stream.of(
+            Arguments.of("Valid without partition", "arn::s3:us-east-1:12345678910:myresource"),
+            Arguments.of("Valid without service", "arn:aws::us-east-1:12345678910:myresource")
+        );
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("validArnTestCases")
+    public void optionalArnFromString_ValidArns_ReturnsPopulatedOptional(String testName, String arnString) {
+        Optional<Arn> optionalArn = Arn.tryFromString(arnString);
+
+        assertThat(optionalArn).isPresent();
+
+        Arn expectedArn = Arn.fromString(arnString);
+        Arn actualArn = optionalArn.get();
+
+        assertThat(actualArn.partition()).isEqualTo(expectedArn.partition());
+        assertThat(actualArn.service()).isEqualTo(expectedArn.service());
+        assertThat(actualArn.region()).isEqualTo(expectedArn.region());
+        assertThat(actualArn.accountId()).isEqualTo(expectedArn.accountId());
+        assertThat(actualArn.resourceAsString()).isEqualTo(expectedArn.resourceAsString());
+
+        assertThat(actualArn.toString()).isEqualTo(arnString);
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("invalidArnTestCases")
+    public void optionalArnFromString_InvalidArns_ReturnsEmptyOptional(String testName, String arnString) {
+        Optional<Arn> optionalArn = Arn.tryFromString(arnString);
+        assertThat(optionalArn).isEmpty();
+    }
+
+    @ParameterizedTest(name = "{0}")
+    @MethodSource("exceptionThrowingArnTestCases")
+    public void tryFromString_InvalidArns_ShouldThrowExceptions(String testName, String arnString) {
+        assertThrows(IllegalArgumentException.class, () -> {
+            Arn.tryFromString(arnString);
+        });
+    }
```

Comment:
Hmm, I thought it would no longer throw exception, did I miss anything?

Reply:
I agree, but I matched the behavior to the behavior that backported to v1.
I think the original implementation was when `service` or `partition` are not provided, we don't explicitly throw an exception within this function. Instead we try to build the arn using the Arn.Builder and use `Validate.paramNotBlank` in the constructor https://github.com/aws/aws-sdk-java-v2/blob/master/core/arns/src/main/java/software/amazon/awssdk/arns/Arn.java#L84-L85

One could argue this is valid validation rather than the same type of "exceptions used as control flow" functionality that currently exists in `fromString`. I don't feel strongly about it either way. Happy to change it.